docs/coding-conventions.md

```diff
@@ -0,0 +1,298 @@
+# Coding conventions
+
+This document outlines a collection of guidelines, style suggestions, and tips
+for writing code in the KubeVirt project. It is partially based on [Kubernetes
+Coding Conventions](https://github.com/kubernetes/community/blob/master/contributors/guide/coding-conventions.md).
+
+The coding conventions in this document are mainly focused on Golang,
+the language used primarily in the KubeVirt project.
+
+Consider each convention a broad and opinionated statement. That means
+maintainers are allowed to make well-motivated exceptions, but it should not
+be the norm.
+
+As a developer you should at least be familiar with the [short coding
+conventions (TLDR)](#short-coding-conventions-tldr) of the KubeVirt project.
+
+# Overview of the most important conventions
+
+When contributing to the KubeVirt project, pay special attention to the
+following:
+
+## Write elegant, cohesive and easily readable code
+
+- If reviewers ask questions about why the code is the way it is, that's a
+  sign that your code is not clear enough.
+- Try improving your code on the expense of writing comments. A properly-named
+  function is better than a comment on a code stanza.
+- Add comments where code-documentation is not enough.
+
+## Avoid nesting and complexity by using early returns
+
+- Deeply nested if/else statements make it harder to understand code.
+- By using early returns you can avoid nesting of if/else statements.
+- Code that is easier to understand will be easier to review and less
+  likely to contain hidden bugs (e.g. control flow issues).
+
+### Bad example
+
+```go
+val, err := doSomething()
+if err == nil {
+    if val {
+        return doSomethingElse()
+    } else {
+        return doAnotherThing()
+    }
+} else {
+    return err
+}
+```
+
+### Good example
+
+```go
+val, err := doSomething()
+if err != nil {
+    return err
+}
+
+if val {
+    return doSomethingElse()
+}
+
+return doAnotherThing()
+```
+
+## Use the same coding style throughout a file
+
+- When adding to a file, stick to the existing coding style to make reviews
```

Comment:
Do you have a better example at hand? You're right about not allowing improvement. I wanted to express, that a PR should not (or not too much, at least separate commits) mix code cleanups and new features.

Reply:
What do you mean by a better example? I'm not sure what the examples below try to illustrate.

Not mixing cleanup with logical changes, then it can be expressed in text, doing this in code examples is hard. Maybe a blog can be written and ref from here.

Keeping the coding style is important, but I would soften it a bit so it will not be used as a reason to reject something better just because the whole file looks bad. Asking the author to fix all the file is not reasonable.

Maybe something like:
> Having a consistent coding style thought a file is recommended. In case a cleaner agreed style is suggested by new code, an attempt should be made to adjust the existing code as a follow up contribution. 

And what you mentioned about separation between style and business logic changes, makes total sense.